### DISCUSSION TITLE
Initial SPA App loading - Blank Screen - needs loading indicator for better UX

### DISCUSSION THREAD

**Discussion by sonicviz** [quasarframework/quasar#17323] [General - Components / Directives / etc]
Hi, 

I've written an SPA webapp, and optimized it as much as I can for now with a variety of techniques. 
Still not as fast as I would like but it'll do for now. 

However, the loading UX is pretty bad as it just presents a blank white screen the whole time for the initial load, which is a pretty terrible UX. 

I use the loading spinner in the app in a few places, but this is a different situation and there's no guidance on how or even if you can use it for an initial app loading context. 

Is there an elegant solution to this, to show a "Please wait a moment, app is loading" message with a spinner while the main app loads, then dismisses it to present the main app? Can't find any existing discussions on this topic. 

Thanks, 
Paul

**Comment by sonicviz** [comment]
I devised a solution using a Loading message in the index template, and removing it using onMounted in the layouts. 
It works, better than a blank screen which is a terrible UX, but could be better.

Open to suggestions!

**Reply by stefanvanherwijnen** [reply]
You should keep your MainLayout minimal and lazy load the content. If the initial load of the MainLayout is too slow for your needs, you either can use the Loading Plugin like you did, or convert your SPA to a SSR app (which makes the server do the heavy lifting).

**Reply by sonicviz** [reply]
Are there any good working examples showing that in action?
Atm I've managed to get it too "borderline" acceptable, but haven't seen any good lazy load examples, or SPA to SSR conversions. 
I think there were some issues to do with SSR, I'll have to check my notes. iirc it's not for all scenarios.

**Reply by stefanvanherwijnen** [reply]
The main thing you should keep in mind and read is https://quasar.dev/quasar-cli-vite/lazy-loading.

With regards to SSR, it should be applicable to all scenarios, but it can be a pain in the ass to actually get it working. Especially if you want to render dynamic data from a database on the server it gets really complicated. My solution is to load all dynamic data with an API call in the onMounted hook, such that the server only has to render the layout/page.

Unless you have an extremely slow internet connection, a Quasar SPA should still load/show within e.g. a second if you implement lazy-loading correctly. You should also check your dependencies/bundle size with [rollup-plugin-visualizer](https://www.npmjs.com/package/rollup-plugin-visualizer) for example. Or use [Lighthouse](https://www.npmjs.com/package/lighthouse) to check what is actually taking so long to load.

**Reply by sonicviz** [reply]
OK, thanks. I'll take another look into it.

**Reply by sonicviz** [reply]
Actually I am using Lazy Loading already, forgot!
Desktop is ok
![image](https://github.com/quasarframework/quasar/assets/200370/1c7c72ff-625a-4b04-8ed3-5340ef4c0a2c)
But the mobile has a couple of issues (long LCP) I can't figure out
![image](https://github.com/quasarframework/quasar/assets/200370/676568e0-9bae-45aa-8714-ace7cf4636be)

*reminds self to fix the accessibility, seeing as I have a certificate in it. Should have been more on the ball with that.

**Reply by stefanvanherwijnen** [reply]
You should be able to get more details on the LCP in Lighthouse to find the difference between desktop and mobile. If loading is still too slow with this Lighthouse score I'd suggest you invest in a new internet connection or (file) server :grin: .

**Reply by sonicviz** [reply]
I don't think that's the issue (vercel for FE), but the backend is still in beta (Appwrite cloud, though I don't think that's the cause here). 
Still, I'll keep poking at it. I'm a little suspect of Lighthouse sometimes, it visually shows much faster than the timings indicate. 
That said, mobile is 3G fast setting, which isn't the fastest now days for most urban environments. Desktop speed is in acceptable bounds.

Thanks for the sounding board feedback though, helpful to do!
